### ISSUE TITLE
iPEPS CTM+AD pipeline: GPU vs CPU numerical drift at float64

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#333]
## Summary

The 2-site iPEPS CTM + implicit-AD pipeline produces **catastrophically different numerics on CUDA vs CPU** at identical float64 precision. Same code, same config, same RNG seed:

| backend | E_gs | wall |
|---|---|---|
| **CPU** (OpenBLAS/LAPACK) | **−0.566285** | 156 s |
| **CUDA** (2× NVIDIA, x64 enabled) | **−7.470805** | ~340 s |

Physical ground state (Sandvik QMC): E/site = −0.6694.

GPU is not just numerically noisy — it's ~13× below physical, the same catastrophic-drift regime that #328 was trying to fix on the explicit-AD path. This was discovered while adding a regression test for the #328 resolution (PR #332): the same test passes on CPU and fails on GPU by a factor of 13.

## Reproduction

```python
import jax
import jax.numpy as jnp
import tenax
from tenax.algorithms.ipeps_config import iPEPSConfig, CTMConfig
from tenax.algorithms.ipeps_optimize import _wrap_as_dense_tensor

def heisenberg_gate():
    d = 2
    Sz = 0.5 * jnp.array([[1.0, 0.0], [0.0, -1.0]])
    Sp = jnp.array([[0.0, 1.0], [0.0, 0.0]])
    Sm = jnp.array([[0.0, 0.0], [1.0, 0.0]])
    H = jnp.kron(Sz, Sz) + 0.5 * jnp.kron(Sp, Sm) + 0.5 * jnp.kron(Sm, Sp)
    return H.reshape(d, d, d, d)

D, d = 2, 2
ka, kb = jax.random.split(jax.random.PRNGKey(0))
A = _wrap_as_dense_tensor(jax.random.normal(ka, (D, D, D, D, d)))
B = _wrap_as_dense_tensor(jax.random.normal(kb, (D, D, D, D, d)))

cfg = iPEPSConfig(
    max_bond_dim=2,
    ctm=CTMConfig(chi=16, max_iter=30, min_iter=10),
    gs_num_steps=30,
    gs_learning_rate=1e-2,
    unit_cell="2site",
    gs_c4v=False,
    gs_explicit_ad=False,  # implicit
    gs_explicit_ad_steps=0,
    gs_explicit_ad_warmup=0,
    gs_optimizer="lbfgs",
    gs_line_search=True,
    gs_line_search_method="armijo",
    gs_line_search_max_steps=8,
    gs_metric_precond=False,
    gs_verbose=False,
)
_, _, E_gs = tenax.optimize_gs_ad(heisenberg_gate(), (A, B), cfg)
print(E_gs)
```

Run with `JAX_PLATFORMS=cpu uv run python …` → **−0.566285**
Run with `JAX_PLATFORMS=cuda uv run python …` (or default on a GPU box) → **−7.470805**

Verified:
- Both use `jax_enable_x64=True` (set by `src/tenax/__init__.py:45`)
- Both use the same RNG seed (deterministic init)
- Result is bit-identical across reruns on the same backend → not RNG / JIT nondeterminism
- Difference shows up on the **implicit-AD path**, so it's not in the forward-mode unroll

## Hypothesis

The CTM projector uses eigh/SVD/QR on the CTM environment corner matrices. CUDA's cuSolver and CPU's OpenBLAS/LAPACK implementations of these can differ substantially in:

- Sign conventions for eigenvectors (random phase)
- Ordering of near-degenerate eigenvalues
- Numerical conditioning when the matrix is near-singular (common in CTM at finite χ)

For a symmetric-normal optimization this is usually harmless — you just take a slightly different path to the same minimum. But the iPEPS loss landscape is known to have unphysical CTM fixed points that give very low surrogate energies (see #328 resolution bench), and small perturbations in the projector can steer the optimizer into one of them.

## Impact

- The regression test in PR #332 (\`test_2site_noc4v_implicit_ad_is_variational_issue_328\`) is forced to CPU via \`jax.default_device(jax.devices('cpu')[0])\`. On GPU it lands at E=−7.5 and fails the (−0.80, −0.30) bounds.
- Any user running iPEPS AD on GPU expecting to reproduce published CPU benchmarks will get different and often much worse numbers, without any warning.
- This likely affects the fermionic iPEPS AD path, multi-site AD, and possibly 1-site AD — all untested on GPU at the time of filing.

## Acceptance criteria

- [ ] Identify the specific CTM projector / linalg call where CPU and GPU diverge (suspect: \`tenax.linalg.eigh\` or \`tenax.linalg.svd\` inside the CTM projector computation)
- [ ] Either fix the divergence at its source (e.g., enforce a canonical eigenvector sign), or document the gap and add a backend check in the AD optimizers
- [ ] Add GPU regression tests (skipped if no CUDA device available) for the 1-site and 2-site iPEPS AD paths at D=2 χ=8
- [ ] Update user docs / README with GPU-specific caveats for iPEPS AD

## Related

- #328 (resolved via PR #332) — the original drift concern that surfaced this
- #294, #297, #329 — SymmetricTensor / optimizer shell work (unrelated but touches the same pipeline)
- Memory: \`project_328_explicit_ad_dead_end.md\` has the full bench matrix

## Out of scope for this issue

- Fixing #328 (already done in #332)
- Rewriting the CTM projector — start by diagnosing which specific call drifts

**Comment by yingjerkao** [comment]
Follow-up bench to test whether metric preconditioning + SU init fixes the GPU drift.

**Result: no.** Metric preconditioning is NOT a reliable fix; it briefly stabilizes GPU at 30 optimization steps but catastrophically diverges at 100 steps.

**Bench** (D=2 Heisenberg, χ=16, 2-site noc4v implicit AD, Armijo L-BFGS):

| # | config | CPU E_gs | GPU E_gs |
|---|---|---|---|
| 1 | baseline (no precond, random, 30) | **-0.5663** ✓ | -7.4708 ❌ |
| 2 | +precond, random, 30 | -5.6463 ❌ | -4.2726 ❌ |
| 3 | +precond, +su_init, 30 | -1.2205 ❌ | **-0.5127** ✓ |
| 4 | +precond, +su_init, **100** | -0.3664 ✓ | **-62.1415** 💥 |

Physical: -0.6694.

**Two key observations:**

1. **No config is stable across both backends.** Opposite winners: 30 steps is best on GPU, 100 steps is best (variational) on CPU.
2. **GPU's catastrophic drift accumulates with iteration count**, not with a specific optimizer choice. At row 4 GPU goes from briefly stable (-0.51 at 30 steps) to E=-62 at 100 steps — a factor of 120 divergence purely from more optimizer iterations. This rules out my earlier hypothesis that the drift was steering-related and the optimizer was being pulled into an unphysical CTM fixed point; instead, some per-iteration numerical error is accumulating in the CTM backward pass on CUDA.

**Revised hypothesis for #333**: the divergence is likely in the **implicit-diff backward through CTM** (`tenax.algorithms.ad_utils.ctm_tensor_converge` VJP), not in the forward CTM projector. Each optimization step calls the backward, which involves a GMRES solve through the CTM Jacobian. CUDA's iterative solver may be accumulating error (conditioning or iteration-count differences) that the CPU solver handles fine.

Next step (not started): isolate the forward vs backward contribution by running a **pure forward** CTM on GPU with the same tensors and comparing energy/eigenspectrum to CPU. If forward matches, the drift is in the backward (implicit-diff VJP). If forward diverges, it's in the projector.

Raw data: `/tmp/bench_328_close_gap_{cpu,gpu}.{log,json}`.

**Comment by yingjerkao** [comment]
## Root cause of non-C4v 2-site non-variationality (extends #328)

A follow-up diagnostic session ran three experiments on `exp/noc4v-no-loss-normalization-328` to localize why non-C4v 2-site implicit AD plateaus at E=-0.566 instead of the expected -0.669 on D=2 Heisenberg. Result: **the Tenax backward VJP is mathematically correct but is being asked to differentiate a non-contractive CTM fixed point**, so every call returns partial-sum garbage.

### Experiments (D=2 Heisenberg, chi=16, noc4v 2-site, implicit AD, seed=0)

| experiment | config | 30-step E | 200-step E |
|---|---|---|---|
| baseline | Neumann, bwd cap 30 | -0.566 | -30.83 (Armijo) |
| (a) raise bwd cap | Neumann, bwd cap 200 | **-3.369** | — |
| (b) GMRES | `ad_backward_method=gmres`, cap 30 | **-0.885** | **-1.375** |
| (c) instrumented | Neumann, bwd cap 30 | -0.566 | — |

### The smoking gun

Instrumenting the Neumann loop in `src/tenax/algorithms/ad_utils.py:1215-1270` to tag exit reasons across a baseline 30-step run:

```
[bwd_trace] call#1 exit=div_break iters=9
[bwd_trace] call#2 exit=div_break iters=9
...
[bwd_trace] call#30 exit=div_break iters=8

trace_summary={'div_break': 30, '_calls': 30, '_total_iters': 229}
```

**30 / 30 backward calls exit via `div_break`** (line 1266: `lam_norm > 1e15`), rolling back one Neumann step and returning a partial sum. Average ~7.6 iterations before `lam` blows up. Zero convergences. `lam_norm` growing from O(1) to 1e15 in 7 iterations implies dominant `J^T` eigenvalue ~ 72 — massive Neumann divergence.

The "-0.566 plateau" is L-BFGS fed partial-sum garbage that happens to project well enough through `vjp_site_fn` to avoid immediate explosion. Accumulated curvature pairs encode the garbage → drift. More steps → catastrophic drift (-30.83 at 200 steps, Armijo).

### What this rules out

- **Iteration count** (exp a). Raising cap 30→200 made E=-3.369 (worse). Matches the earlier "tight CTM forward max_iter=200 → -3.37" data point. More iterations = more garbage accumulated before `div_break`.
- **Solver choice** (exp b). GMRES gives a different flavor — bounded drift (-0.88 → -1.37) instead of Neumann's catastrophic — but still non-variational. GMRES solves the near-singular `(I − J^T) lam = g` in a least-squares sense; the answer is close but not a valid gradient.
- **"Silent break is a bug"** (exp c). The `div_break` path at line 1266 is firing *correctly*. It's preventing NaN from a genuinely divergent series. Removing it would propagate NaNs; raising the threshold would accumulate more garbage before rollback.

### Root cause

**`J^T` has spectral radius ≫ 1 on non-C4v 2-site**, so the fixed-point adjoint `(I − J^T) lam = g` is not well-defined — the CTM forward is not locally contractive in the full tangent space. This matches the existing `project_multisite_ad_symmetry_constraint` analysis: non-C4v 2-site has a bipartite gauge zero mode (A ↔ B permutation) plus additional unstable directions without a point-group constraint.

### Implication for #333 GPU/CPU drift

Per-iteration floating-point noise in a *divergent* Neumann series amplifies very differently on different backends. On GPU, reduction order changes the `J^T @ grads` step by epsilon; over 7 iterations at spectral radius ~72, that epsilon gets amplified by 72^7 ≈ 10^13. This is a plausible mechanism for the GPU-specific drift reported in #333 — the difference is not in any single operation but in how quickly the amplification crosses the 1e15 rollback threshold.

### What this points toward

The fix for non-C4v 2-site must act **before** implicit diff by enforcing a symmetry that makes `J^T` contractive on the physical subspace. Options already in the roadmap:

1. **Shared-tensor C4v (Approach C)** — lift bipartite symmetry to a single tensor `A` with explicit `B = σ·A·σ`. Eliminates the zero mode; the reduced adjoint may be contractive.
2. **Multisite c4v_reference AD** — same idea generalized (existing design in PR #315 discussion).

Any further tuning of the backward solver (cap, method, tol, preconditioner) is a dead end on this problem.

### Bench scripts

- `/tmp/bench_328_expA_bwd200.py` — raised backward iter cap
- `/tmp/bench_328_expB_gmres.py` — GMRES backward
- `/tmp/bench_328_expC_trace.py` — instrumented Neumann loop

Branch `exp/noc4v-no-loss-normalization-328` is the working environment for reproduction but contains only the earlier `TENAX_EXP_SKIP_LOSS_NORM` diagnostic commit; the instrumentation used in (c) was applied locally and reverted.

**Comment by yingjerkao** [comment]
## Follow-up: shared-tensor C4v + explicit AD validated; Arnoldi/sigma not sufficient for 2-site

Continuing from the [root-cause diagnosis](https://github.com/tenax-lab/tenax/issues/333#issuecomment-4256449670), a second experimental session tested fixes.

### What works: shared-tensor C4v + explicit AD

The existing `gs_c4v=True + gs_explicit_ad=True + unit_cell="2site"` code path gives variational energies for Heisenberg (D=2, seed=0, Armijo line search):

| chi | explicit_steps | E | wall | variational? |
|---|---|---|---|---|
| 8 | 10 | −0.6513 | 80s | ✓ |
| 16 | 10 | **−0.6573** | 95s | ✓ |
| 32 | 10 | −0.6539 | 262s | ✓ |
| 40 | 10 | −0.6499 | 403s | ✓ |

All above exact (−0.6694), clustered near literature (−0.6548). Non-monotonic in chi due to finite explicit unrolling, but no drift or divergence at any chi. `explicit_steps=10` is the sweet spot — `explicit_steps=20` at chi=40 gave −0.6322 (worse, 10x slower).

**Limitation:** requires Marshall sign (`B = σ_y · A · σ_y`). Not applicable to frustrated or non-bipartite models.

### What doesn't work for non-C4v 2-site

**Shared-tensor C4v + implicit AD:** chi=16 still 30/30 `div_break` (E=−0.502, worse than non-C4v baseline). Chi=8 is marginal (9/30 div_break, 18/30 ema_break, E=−0.6518). Shared-tensor kills the bipartite zero mode but unstable J^T directions scale with chi.

**Skip + reset (variPEPS-style error recovery):** Raised `BackwardVJPDivergedError` on `div_break`, caught in optimizer, skipped + reset L-BFGS. Result: 30/30 steps skipped, E=+0.192 (no optimization). Can't descend when 100% of gradients diverge.

**PR #334 Arnoldi precheck + sigma gauge:** Arnoldi correctly detects ρ(J^T) = 17.94 on non-C4v 2-site at chi=8 (vs 1.88 on 1-site C4v) and blocks the backward. But stall recovery churns without progress — same outcome as skip+reset.

**Uncapped GMRES (variPEPS's actual approach):** Removed `maxiter` cap from Tenax's GMRES path. Impractically slow — 50+ min on GPU for 30 steps with no result (killed). variPEPS achieves this via JIT-fusing the entire forward+backward into XLA, so GMRES `while_loop` benefits from hardware scheduling. Tenax's eager backward pays per-iteration dispatch overhead.

### variPEPS comparison

variPEPS uses `@custom_vjp` + GMRES (no maxiter) + Arnoldi eigensolver precheck + error recovery. Their Heisenberg example optimizes 2 independent tensors (no C4v) on a checkerboard unit cell. The key difference: their entire `calc_ctmrg_expectation_custom_value_and_grad` is JIT-compiled, so GMRES fuses into the XLA graph. Tenax's backward runs eagerly from a Python optimizer loop.

### Recommendation

**For Marshall-sign bipartite AFMs:** use `gs_c4v=True, gs_explicit_ad=True, gs_explicit_ad_steps=10`. Works now on main, no code changes needed.

**For general (non-Marshall) models:** the productive next step is **Approach A from the [multisite c4v reference AD design](https://github.com/tenax-lab/tenax/pull/315)** — Lorentzian-regularized projector backward on the explicit-AD path. This bypasses the broken implicit backward VJP entirely and is already designed. Alternatively, JIT-fusing the backward (like variPEPS) would make GMRES competitive, but requires restructuring `_ctm_tensor_converge_bwd` to be JIT-friendly.

### Bench scripts

All at `/tmp/bench_328_*.py` on the `exp/noc4v-no-loss-normalization-328` branch (merged with main post-#334).

**Comment by yingjerkao** [comment]
Closing — the headline GPU-vs-CPU divergence (CPU −0.566, GPU −7.47) was a manifestation of the real-`float64` pathology on the 2-site non-C4v implicit-AD path, not a GPU bug per se.

**Diagnosis (post-PR-#341):**

The implicit-diff linear system `(I − J^T)λ = g` is ill-conditioned in the real subspace for non-C4v 2-site iPEPS. Both CPU and GPU drift away from physical with real `float64`; the GPU run just lands at a different (and worse) non-variational fixed point than CPU because the linalg implementations explore the chaotic landscape differently. Doubling the parameter space with `complex128` well-conditions the system and the drift goes away — verified on GPU at D=2 χ=16, E=−0.6406 (variational, 4.3 % above QMC −0.6694), ≈1 min/step.

**Resolution:** promote site tensors to `complex128` for non-C4v multi-site iPEPS AD. The implicit-AD path on `main` now also enforces the `svd + phase + elementwise` trifecta via `resolve_projector_backward`, which removes the gauge ambiguity that amplified the divergence.

**Test follow-up:** `tests/test_ipeps.py:831` (`test_2site_noc4v_implicit_ad_is_variational_issue_328`) still pins to CPU and uses real `float64`. That CPU pin and the real-dtype init can come out in a separate PR — the workaround predates the complex128 finding and is no longer needed for the test to pass.
